### PR TITLE
Capacityfactors for onshore and offshore wind and pv

### PR DESCRIPTION
**Pull request by jnnr**
Renewables

**Uses these scripts**
```
scripts/wind-and-solar/shared_coast.py
scripts/wind-and-solar/capacityfactors.py
scripts/wind-and-solar/capacityfactors_offshore.py
```
**Uses these rules**
_shapes.smk:_
```
rule download_eez
rule eez
```
_wind-and-solar.smk:_
```
rule shared_coast
rule download_capacity_factors_wind_and_solar
rule capacity_factors_wind_and_solar
rule capacity_factors_offshore
```
**Produces these outputs:**
```
build/data/eez.geojson
build/data/{resolution}/shared-coast.csv
build/models/{resolution}/timeseries/supply/capacityfactors-wind-offshore.csv
build/models/{resolution}/timeseries/supply/capacityfactors-{technology}.csv
```

### CODE REVIEW

File: modules/capacity_factors_wind_pv/workflow/internal/internal_config.yaml
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/internal/internal_config.yaml:3]: Looking at `config/default.yaml`, I think there are a bunch of things you want to pass here to protect them from user error (and to avoid having to define a `schema` for them).

Good candidates:
- Downloaded datasources (eez, cap-fact).
- The spatial scope (unless those datasources are global, in which case it makes no sense to have a spatial scope).
- The min-max capacity range.
- The spatial 'bounds' (again, if the datasource is global, this could be removed).
- The thresholds

File: modules/capacity_factors_wind_pv/workflow/rules/downloads.smk
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/rules/downloads.smk:15]: Please add a check to deactivate this (see https://github.com/calliope-project/ec_modules/pull/39)

File: modules/capacity_factors_wind_pv/workflow/Snakefile
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/Snakefile:14]: Please add a fully restrictive schema for user config!
Otherwise, users have no way of knowing if they are configuring this module correctly.

File: modules/capacity_factors_wind_pv/workflow/schemas/config.schema.yaml
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/schemas/config.schema.yaml:1]: This should fully check user config.

Thanks to internal config, I think the list of stuff will actually be rather small.

File: modules/capacity_factors_wind_pv/workflow/scripts/.gitkeep
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/scripts/.gitkeep:1]: Feel free to delete :)

File: modules/capacity_factors_wind_pv/config/default.yaml
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/config/default.yaml:55]: To make the workflow much more flexible, try changing things to use a wildcard (`{resolution}/{year}`), removing the need for this configuration.
It might be necessary to check your sourced datasets for the years they enable, and set a global wildcard constraint for it.

```
wildcard_constraint:
    year = "|".join([str(i) for i in range(2000, 2019)])
```

File: modules/capacity_factors_wind_pv/workflow/Snakefile
Comment (irm-codebase) [inline modules/capacity_factors_wind_pv/workflow/Snakefile:50]: In the new IO, the all rule is deactivated via `use_default_user_config: False`.
If enabled, you should run a user input example.

**Review by irm-codebase — CHANGES_REQUESTED** [review]
Overall pretty good! Unfortunately I came in late with the IO standard, so a few changes are necessary.
Beyond that:
- Please use the schema
- Please use the internal configuration
- Please add a bit more info to the README.

See specific comments below.

File: modules/capacity_factors_wind_pv/workflow/internal/internal_config.yaml
Comment (jnnr) [reply modules/capacity_factors_wind_pv/workflow/internal/internal_config.yaml:3]: Done.

File: modules/capacity_factors_wind_pv/config/default.yaml
Comment (jnnr) [reply modules/capacity_factors_wind_pv/config/default.yaml:9]: Will harmonize that soon, also in the other new modules, but first I'll merge.

**Comment by jnnr** [comment]
As you did not approve this one yet, I ask for another quick review after adopting your changes. I did not implement the wildcard suggestion yet. It should be adopted for demand_electricity and biofuels too - can be done later, maybe? Feel free to merge this branch here if you are happy with it.

File: modules/capacity_factors_wind_pv/config/default.yaml
Comment (irm-codebase) [reply modules/capacity_factors_wind_pv/config/default.yaml:9]: Agreed. This can wait for later.

**Comment by irm-codebase** [comment]
@jnnr looks good overall. The only thing that slipped was the `AUTHORS` file and the LICENSE, which need small updated. I'll make a small update and merge this in.

**Review by irm-codebase — APPROVED** [review]

**Pull request merged**
